File: app/src/main/java/de/bikebean/app/db/settings/settings/number_settings/WifiAccessPoints.java
package de.bikebean.app.db.settings.settings.number_settings;

import androidx.annotation.NonNull;

import java.util.ArrayList;

import de.bikebean.app.db.settings.settings.NumberSetting;
import de.bikebean.app.db.settings.settings.WappState;
import de.bikebean.app.db.sms.Sms;
import de.bikebean.app.db.state.State;
import de.bikebean.app.db.state.StateFactory;

public class WifiAccessPoints extends NumberSetting {

    private static final @NonNull State.KEY key = State.KEY.WIFI_ACCESS_POINTS;
    private static final @NonNull State.KEY numberKey = State.KEY.NO_WIFI_ACCESS_POINTS;

    public WifiAccessPoints(final @NonNull String wifiAccessPoints, final @NonNull Sms sms) {
        super(wifiAccessPoints, sms, key,
                StateFactory.createNumberState(
                        sms, numberKey,
                        getNumberFromString(wifiAccessPoints), State.STATUS.CONFIRMED
                ),
                new WifiAccessPointList(wifiAccessPoints.split("\n"))
        );
    }

    public WifiAccessPoints(final @NonNull Sms sms,
                            final @NonNull WifiAccessPointsGetter wifiAccessPointsGetter) {
        this(wifiAccessPointsGetter.getWifiAccessPoints(), sms);
    }

    public WifiAccessPoints(final @NonNull WappState wappState) {
        this(wappState.getWifiAccessPoints().getLongValue(), wappState.getSms());
    }

    public WifiAccessPoints() {
        super("", key, StateFactory.createNullState(),
                new WifiAccessPointList(new String[]{})
        );
    }

    public interface WifiAccessPointsGetter {
        @NonNull String getWifiAccessPoints();
    }

    private static int getNumberFromString(final @NonNull String string) {
        if (string.isEmpty())
            return 0;
        else return string.split("\n").length;
    }

    public static class WifiAccessPointList extends ArrayList<WifiAccessPoint> {
        WifiAccessPointList(final @NonNull String[] stringArrayWapp) {
            parse(stringArrayWapp);
        }

        private void parse(final @NonNull String[] stringArrayWapp) {
            for (final @NonNull String s : stringArrayWapp)
                if (!s.equals("    ") && !s.isEmpty()) {
                    // Länge des Substrings ist Unterscheidungskriterium
                    final @NonNull WifiAccessPoint wap = new WifiAccessPoint();
                    wap.macAddress = s.substring(2);
                    wap.signalStrength = Integer.parseInt("-" + s.substring(0, 2));
                    wap.toMacAddress();

                    this.add(wap);
                }
        }
    }

    public static class WifiAccessPoint extends RawNumberSettings {
        public String macAddress;
        public Integer signalStrength;

        void toMacAddress() {
            @NonNull String str = macAddress;
            final @NonNull StringBuilder tmp = new StringBuilder();
            final int divisor = 2;

            while (str.length() > 0) {
                String nextChunk = str.substring(0, divisor);
                tmp.append(nextChunk);
                if (str.length() > 2)
                    tmp.append(":");

                str = str.substring(divisor);
            }

            macAddress = tmp.toString();
        }
    }
}

File: app/src/main/java/de/bikebean/app/db/settings/settings/add_to_list_settings/number_settings/WifiAccessPoints.kt
package de.bikebean.app.db.settings.settings.add_to_list_settings.number_settings

import de.bikebean.app.db.settings.settings.add_to_list_settings.NumberSetting
import de.bikebean.app.db.settings.settings.add_to_list_settings.WappState
import de.bikebean.app.db.sms.Sms
import de.bikebean.app.db.state.State
import de.bikebean.app.db.state.State.KEY
import de.bikebean.app.db.state.StateFactory
import kotlinx.serialization.Serializable

class WifiAccessPoints : NumberSetting {

    override val list: List<WifiAccessPoint>

    constructor(sms: Sms, wifiAccessPoints: String) :
            super(wifiAccessPoints, sms, key,
                    StateFactory.createNumberState(
                            sms, numberKey,
                            getNumberFromString(wifiAccessPoints).toDouble(), State.STATUS.CONFIRMED
                    )
            ) {
        list = WifiAccessPointListBuilder(wifiAccessPoints)
            }

    constructor(wappState: WappState) : this(wappState.sms, wappState.wifiAccessPoints.longValue)

    constructor() : super("", key, StateFactory.createNullState()) {
        list = WifiAccessPointListBuilder("")
    }

    class WifiAccessPointListBuilder(stringArrayWapp: String) : ArrayList<WifiAccessPoint>() {

        init {
            if (stringArrayWapp.isNotEmpty())
                stringArrayWapp.split("\n").forEach { s ->
                    add(WifiAccessPointBuilder(s).get())
                }
        }

        fun add(element: WifiAccessPoint?): Boolean {
            return when (element) {
                null -> false
                else -> super.add(element)
            }
        }

    }

    @Serializable
    data class WifiAccessPoint(
            val macAddress: String,
            val signalStrength: Int)
        : RawNumberSettings()

    class WifiAccessPointBuilder(s: String) {

        private val wifiAccessPoint: WifiAccessPoint? = when {
            s != "    " && s.isNotEmpty() ->
                WifiAccessPoint(
                        createMacAddress(s.substring(2)),
                        ("-" + s.substring(0, 2)).toInt())
            else -> null
        }

        fun get(): WifiAccessPoint? = wifiAccessPoint


        private fun createMacAddress(rawMacAddress: String): String = StringBuilder().let {
            splitIntoChunksRecursive(rawMacAddress, it)
            it.toString()
        }

        private fun splitIntoChunksRecursive(remainder: String, final: StringBuilder) : Unit =
                final.let {
                    it.append(remainder.substring(0 until 2))

                    if (remainder.length <= 2) return

                    it.append(":")
                    splitIntoChunksRecursive(remainder.substring(2), it)
                }
    }

    companion object {
        private val key = KEY.WIFI_ACCESS_POINTS
        private val numberKey = KEY.NO_WIFI_ACCESS_POINTS
    }
}